services/brig/src/Brig/Data/UserPendingActivation.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2020 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Brig.Data.UserPendingActivation
+  ( trackExpiration,
+    getAllTrackedExpirations,
+    removeTrackedExpiration,
+    removeTrackedExpirations,
+    UserPendingActivation (..),
+  )
+where
+
+import Brig.App (AppIO)
+import Cassandra
+import Data.Id (UserId)
+import Data.Time (UTCTime)
+import Imports
+
+data UserPendingActivation = UserPendingActivation
+  { upaUserId :: !UserId,
+    upaDay :: !UTCTime
+  }
+  deriving stock (Eq, Show, Ord)
+
+-- | Note: Call this function only after an invitation for the user has been created
```

Comment:
This Note is no longer longer true since we updated the logic (if it ever was true). I'll remove it.

When brig crashes between creating the invitation and making this entry, then the user won't get garbage-collected.

The conditions for the user to get deleted are quite strict now:
- user has entry in table (which is only added after an invitation was created) AND
- after the expiration date the user is in state "PendingActivation"

Reply:
2a64d2825a101c181907417b2df6374035ef7834